src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2062,444 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+									   "%ld and %ld", get_rel_name(relationId),
+									   shardId, shardInterval->shardId)));
+			}
+		}
+
+		table_close(distShardRelation, NoLock);
+	}
+}
+
+
+/*
+ * citus_internal_add_placement_metadata is an internal UDF to
+ * add a row to pg_dist_placement.
+ */
+Datum
+citus_internal_add_placement_metadata(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	int64 shardId = PG_GETARG_INT64(1);
+	char shardState = PG_GETARG_CHAR(2);
+	int64 shardLength = PG_GETARG_INT64(3);
+	int32 groupId = PG_GETARG_INT32(4);
+	int64 placementId = PG_GETARG_INT64(5);
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* only owner of the table is allowed to modify the metadata */
+		EnsureTableOwner(relationId);
+
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		/*
+		 * Even if the table owner is a malicious user, as long as the shard placements
+		 * fit into basic requirements of Citus metadata, the user can only affect its
+		 * own tables. Given that the user is owner of the table, we should allow.
+		 */
+		EnsureShardPlacementMetadataIsSane(shardId, placementId, shardState,
+										   shardLength, groupId);
+	}
+
+	InsertShardPlacementRow(shardId, placementId, shardState, shardLength, groupId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureShardPlacementMetadataIsSane ensures if the input parameters for
+ * the shard placement metadata is sane.
+ */
+static void
+EnsureShardPlacementMetadataIsSane(int64 shardId, int64 placementId,
+								   char shardState, int64 shardLength,
+								   int32 groupId)
+{
+	if (shardId <= INVALID_SHARD_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard id: %ld", shardId)));
+	}
+
+	if (placementId <= INVALID_PLACEMENT_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Shard placement has invalid placement id "
+							   "(%ld) for shard(%ld)", placementId, shardId)));
+	}
+
+	if (!(shardState == SHARD_STATE_ACTIVE ||
+		  shardState == SHARD_STATE_TO_DELETE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard state: %d", shardState)));
+	}
+
+	bool nodeFound = false;
+	bool includeNodesFromOtherClusters = false;
+	List *workerNodeList = ReadDistNode(includeNodesFromOtherClusters);
+	WorkerNode *currentNode = NULL;
+	foreach_ptr(currentNode, workerNodeList)
+	{
+		if (currentNode->groupId == groupId)
+		{
+			nodeFound = true;
+			break;
+		}
+	}
+	if (!nodeFound)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Node with group id %d for shard placement "
+							   "%ld does not exist", groupId, shardId)));
+	}
+
+	List *placementList = BuildShardPlacementList(shardId);
+	GroupShardPlacement *placement = NULL;
+	foreach_ptr(placement, placementList)
+	{
+		if (placement->groupId == groupId)
```

Comment:
can we make this a UNIQUE constraint on (shardid, groupid) instead?

Reply:
added that, which makes sense  in general